nilearn/decoding/decoder.py

```diff
@@ -0,0 +1,799 @@
+"""High-level decoding object that exposes standard classification
+
+and regression strategies such as SVM, LogisticRegression and Ridge,
+with optional feature selection, and integrated parameter selection.
+"""
+# Authors : Yannick Schwartz
+#           Andres Hoyos-Idrobo
+#
+# License: simplified BSD
+
+import sklearn
+import itertools
+import warnings
+import numpy as np
+from distutils.version import LooseVersion
+from sklearn.externals.joblib import Parallel, delayed, Memory
+from sklearn.linear_model.ridge import Ridge, RidgeClassifier, _BaseRidge
+from sklearn.linear_model import LogisticRegression
+from sklearn.svm import LinearSVC, SVR
+from sklearn.preprocessing import LabelBinarizer
+from sklearn.svm.bounds import l1_min_c
+from sklearn.base import RegressorMixin
+from sklearn.linear_model.base import LinearModel
+from sklearn.utils.extmath import safe_sparse_dot
+from sklearn import clone
+
+try:
+    from sklearn.utils import atleast2d_or_csr
+except ImportError: # sklearn 0.15
+    from sklearn.utils import check_array as atleast2d_or_csr
+
+try:
+    from sklearn.grid_search import ParameterGrid
+except ImportError: # sklearn 0.18
+    from sklearn.model_selection import ParameterGrid
+
+from ..input_data.masker_validation import check_embedded_nifti_masker
+from .._utils.param_validation import _adjust_screening_percentile
+from .._utils.fixes import check_scoring
+from .._utils.fixes import check_X_y
+from .._utils.fixes import check_is_fitted
+from .._utils.compat import _basestring
+from .._utils.fixes import check_cv
+from .._utils.param_validation import check_feature_screening
+from .._utils import CacheMixin
+
+
+SUPPORTED_ESTIMATORS = dict(
+    svc_l1=LinearSVC(penalty='l1', dual=False),
+    svc_l2=LinearSVC(penalty='l2'),
+    svc=LinearSVC(penalty='l2'),
+    logistic_l1=LogisticRegression(penalty='l1'),
+    logistic_l2=LogisticRegression(penalty='l2'),
+    logistic=LogisticRegression(penalty='l2'),
+    ridge_classifier=RidgeClassifier(),
+    ridge_regression=Ridge(),
+    ridge=Ridge(),
+    svr=SVR(kernel='linear'),)
+
+
+def _check_param_grid(estimator, X, y, param_grid):
+    """Check param_grid and return sensible default if none is given.
+    """
+    if param_grid is None:
+        param_grid = {}
+        # define loss function
+        if isinstance(estimator, LogisticRegression):
+            loss = 'log'
+        elif isinstance(estimator, (LinearSVC, _BaseRidge, SVR)):
+            loss = 'squared_hinge'
+
+        if LooseVersion(sklearn.__version__) <= LooseVersion('0.17'):
+            if loss == 'squared_hinge':
+                loss = 'l2'
+
+        if hasattr(estimator, 'penalty') and (estimator.penalty == 'l1'):
+            min_c = l1_min_c(X, y, loss=loss)
+        else:
+            min_c = 0.5
```

Comment:
I find the whole set of tests above too open: it is hard to know what estimators are accepted by the code, and which one will work by chance because they go through the chain of if/then without error (aka "unbounded cyclomatic complexity")